### PR TITLE
wip sdcard board

### PR DESCRIPTION
**Pull request by Andrey1994**
Signed-off-by: Andrey Parfenov <a1994ndrey@gmail.com>

Any ideas why I see arrayindexoutofrange and changes in UI are welcome

### CODE REVIEW

**Comment by Andrey1994** [comment]
fixed index error, need to change smth in control panel to set new data source

**Comment by Andrey1994** [comment]
Maybe intead Integer.parseInt we need to use this https://github.com/brainflow-dev/brainflow/blob/master/src/utils/inc/custom_cast.h#L7 (in java of course) but old SDConverter didnt...

File: OpenBCI_GUI/FileBoard.pde [hunk 1]
```diff
@@ -0,0 +1,12 @@
+interface FileBoard {
```
Comment (daniellasry) [inline OpenBCI_GUI/FileBoard.pde:1]: Is the plan to have DataSourcePlayback also implement the FileBoard interface?

File: OpenBCI_GUI/DataSourceSDCard.pde [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    private int samplingRate;
+    private ArrayList<double[]> data;
+    private int[] exgChannels;
+    private int totalChannels;
+    private boolean streaming;
+    private double startTime;
+    private int counter;
+    private int currentSample;
+    private int timeOfLastUpdateMS;
+    private boolean initialized;
```
Comment (daniellasry) [inline OpenBCI_GUI/DataSourceSDCard.pde:13]: is this boolean used for anything?

File: OpenBCI_GUI/DataSourceSDCard.pde [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    }
+
+    private void parseRow(String[] row, int numChannels) {
+        double[] line = new double[totalChannels];
+        if (row.length < numChannels - 1) {
+            return;
+        }
+        for (int i = 0; i < numChannels; i++) {
+            Integer val = Integer.parseInt(row[i], 16);
```
Comment (daniellasry) [inline OpenBCI_GUI/DataSourceSDCard.pde:80]: This looks fine. It should work.

File: OpenBCI_GUI/DataSourceSDCard.pde [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        if (row.length >= numChannels + 4) {
+            accel_x = Integer.parseInt(row[numChannels + 1], 16);
+            accel_y = Integer.parseInt(row[numChannels + 2], 16);
+            accel_z = Integer.parseInt(row[numChannels +3], 16);
+        }
+        line[line.length - 4] = BoardCytonConstants.accelScale * accel_x;
+        line[line.length - 3] = BoardCytonConstants.accelScale * accel_y;
+        line[line.length - 2] = BoardCytonConstants.accelScale * accel_z;
+        // add timestamp
+        double delay = (numChannels == 16) ? (1.0 / 125) : (1.0 / 250);
```
Comment (daniellasry) [inline OpenBCI_GUI/DataSourceSDCard.pde:97]: you could use getSampleRate() here

**Comment by daniellasry** [comment]
I think it looks good so far

**Comment by daniellasry** [comment]
I did a test and indeed it seems liek there is a problem with parsing the data. I tested square wave, and it looks exactly like the issue we had with BDF: https://github.com/OpenBCI/OpenBCI_GUI/pull/737/files

**Comment by daniellasry** [comment]
Try these parsing functions:

```
    private int parseInt24Hex(String hex) {
        if (hex.charAt(0) > '7') {  // if the number is negative	
            hex = "FF" + hex;   // keep it negative	
        } else {                  // if the number is positive	
            hex = "00" + hex;   // keep it positive	
        }

        return unhex(hex);
    }

    private int parseInt16Hex(String hex) {
        if (hex.charAt(0) > '7') {  // if the number is negative	
            hex = "FFFF" + hex;   // keep it negative	
        } else {                  // if the number is positive	
            hex = "0000" + hex;   // keep it positive	
        }

        return unhex(hex);
    }
```

File: OpenBCI_GUI/FileBoard.pde (hunk 1 above)
Comment (Andrey1994) [reply OpenBCI_GUI/FileBoard.pde:1]: I think it does, if not - need to add it

File: OpenBCI_GUI/DataSourceSDCard.pde (hunk 2 above)
Comment (Andrey1994) [reply OpenBCI_GUI/DataSourceSDCard.pde:13]: Probably it was but now not, can remove

File: OpenBCI_GUI/DataSourceSDCard.pde (hunk 4 above)
Comment (Andrey1994) [reply OpenBCI_GUI/DataSourceSDCard.pde:97]: Before for the first sample getSampleRate was set after first parsed package, so I hardcoded it. Now its possible

**Comment by Andrey1994** [comment]
What is unhex? There is no such function in java and I dont see it in gui

**Comment by Andrey1994** [comment]
And it looks strange... its not a conversion from int24/16 I use in brainflow....

File: OpenBCI_GUI/FileBoard.pde (hunk 1 above)
Comment (Andrey1994) [reply OpenBCI_GUI/FileBoard.pde:1]: fixed

**Comment by daniellasry** [comment]
unhex is from processing: https://processing.org/reference/unhex_.html

I don't know why it works and parseInt does not

**Comment by daniellasry** [comment]
Could you change the string "CONVERT SD FOR PLAYBACK" to something like "SELECT SD FOR PLAYBACK"?

Also, could you add a line in CHANGELOG.md, under Improvements, saying that SD files can be played back directly without converting?

**Comment by Andrey1994** [comment]
Its already renamed: https://github.com/OpenBCI/OpenBCI_GUI/pull/800/files#diff-5295e497df226f2138eae5e540267aeaR1066 if there are other places I dont see them

**Comment by daniellasry** [comment]
Yes, ControlPanel.pde line 2549

**Comment by daniellasry** [comment]
Also, WidgetManager.pde line 69, replace that with if(currentBoard instanceof DataSourcePlayback)
We don't want that widget showing in SD card playback.

**Comment by Andrey1994** [comment]
> Also, WidgetManager.pde line 69, replace that with if(currentBoard instanceof DataSourcePlayback)
> We don't want that widget showing in SD card playback.

Why not?

**Comment by Andrey1994** [comment]
Ah, makes sense cause probably this widget set playback file

**Comment by daniellasry** [comment]
One bug: If i select playback file then select SD file, it will run playback file
Maybe you need to clear the playback file name when user selects sd card?

**Pull request merged**